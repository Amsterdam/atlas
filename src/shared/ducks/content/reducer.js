import { initialState, REDUCER_KEY, SET_ITEM } from './constants';

export { REDUCER_KEY as CONTENT };

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case SET_ITEM:
      return {
        ...state,
        item: action.payload
      };

    default:
      return state;
  }
}
