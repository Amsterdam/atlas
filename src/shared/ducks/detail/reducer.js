import { ROUTER_NAMESPACE, routing } from '../../../app/routes';
import { FETCH_DETAIL, initialState, SET_VIEW, SHOW_DETAIL, REDUCER_KEY } from './constants';
import paramsRegistry from '../../../store/params-registry';
import PAGES from '../../../app/pages';

export { REDUCER_KEY as DETAIL };

export default function detailReducer(state = initialState, action) {
  if (action.type &&
    action.type.startsWith(ROUTER_NAMESPACE) &&
    !action.type.includes(PAGES.DATA_DETAIL)
  ) {
    return initialState;
  }

  const enrichedState = {
    ...state,
    ...paramsRegistry.getStateFromQueries(REDUCER_KEY, action)
  };
  switch (action.type) {
    case routing.dataDetail.type: {
      const { id: idString, type, subtype } = action.payload;

      // Package `redux-first-router` doesn't support parameters starting with numbers
      // So prefix `id` is added to numerical IDs and is filtered out here
      const id = idString.startsWith('id') ? idString.substr(2) : idString;

      return {
        ...enrichedState,
        id,
        type,
        subtype
      };
    }

    case FETCH_DETAIL:
      return {
        ...enrichedState,
        display: undefined,
        geometry: undefined,
        isLoading: true
      };

    case SHOW_DETAIL:
      return {
        ...enrichedState,
        display: action.payload.display,
        geometry: action.payload.geometry,
        isLoading: false
      };

    case SET_VIEW:
      return {
        ...enrichedState,
        view: action.payload
      };

    default:
      return enrichedState;
  }
}
