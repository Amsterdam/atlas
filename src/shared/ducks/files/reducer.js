import paramsRegistry from '../../../store/params-registry'
import { initialState, REDUCER_KEY, SET_CURRENT_FILE, SET_CURRENT_TYPE } from './constants'

export { REDUCER_KEY as FILES_REDUCER }

const reducer = (state = initialState, action) => {
  const enrichedState = {
    ...state,
    ...paramsRegistry.getStateFromQueries(REDUCER_KEY, action),
  }
  switch (action.type) {
    case SET_CURRENT_FILE:
      return {
        ...enrichedState,
        fileName: action.payload,
      }

    case SET_CURRENT_TYPE: {
      return {
        ...enrichedState,
        type: action.payload,
      }
    }

    default:
      return enrichedState
  }
}

export default reducer
