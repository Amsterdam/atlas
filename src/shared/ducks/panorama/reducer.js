import { ROUTER_NAMESPACE, routing } from '../../../app/routes';
import { PANORAMA_CONFIG } from '../../services/panorama-api/panorama-api';
import PAGES from '../../../app/pages';
import {
  CLOSE_PANORAMA,
  FETCH_PANORAMA_ERROR,
  FETCH_PANORAMA_REQUEST,
  FETCH_PANORAMA_SUCCESS,
  initialState, SET_PANORAMA_LOCATION,
  SET_PANORAMA_ORIENTATION,
  SET_PANORAMA_VIEW,
  SET_PANORAMA_YEAR
} from './constants';
import { getStateFromQuery } from '../../../store/query-synchronization';
import urlParams from './query';

export { REDUCER_KEY } from './constants';

export default function reducer(state = initialState, action) {
  if (action.type &&
    action.type.startsWith(ROUTER_NAMESPACE) &&
    !action.type.includes(PAGES.PANORAMA)
  ) {
    return initialState;
  }

  const enrichedState = {
    ...state,
    ...getStateFromQuery(urlParams, action)
  };
  switch (action.type) {
    case routing.panorama.type: {
      return {
        ...enrichedState,
        id: action.payload.id
      };
    }

    case FETCH_PANORAMA_REQUEST:
      return {
        ...enrichedState,
        id: action.payload.id,
        isLoading: true
      };

    case SET_PANORAMA_YEAR:
      return {
        ...enrichedState,
        year: action.payload
      };

    case FETCH_PANORAMA_SUCCESS:
      return {
        ...enrichedState,
        date: action.payload.date,
        pitch: state.pitch || initialState.pitch,
        fov: state.fov || PANORAMA_CONFIG.DEFAULT_FOV,
        hotspots: action.payload.hotspots,
        isLoading: false,
        location: action.payload.location,
        image: action.payload.image
      };
    case FETCH_PANORAMA_ERROR:
      return {
        ...enrichedState,
        isLoading: false,
        isError: true // TODO: refactor, show error
      };

    case SET_PANORAMA_ORIENTATION:
      return {
        ...enrichedState,
        heading: action.payload.heading,
        pitch: action.payload.pitch,
        fov: action.payload.fov
      };

    case CLOSE_PANORAMA:
      return {
        ...enrichedState,
        location: initialState.location,
        image: initialState.image
      };

    case SET_PANORAMA_VIEW:
      return {
        ...enrichedState,
        view: action.payload
      };

    case SET_PANORAMA_LOCATION:
      return {
        ...enrichedState,
        location: action.payload
      };

    default:
      return enrichedState;
  }
}

