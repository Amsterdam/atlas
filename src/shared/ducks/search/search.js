import { createSelector } from 'reselect';
import get from 'lodash.get';
import isObject from '../../services/is-object';
import BaseCoder from '../../services/base-coder/base-coder';
import { routing } from '../../../app/routes';

export const REDUCER_KEY = 'search';

export const FETCH_MAP_SEARCH_RESULTS_REQUEST = 'FETCH_MAP_SEARCH_RESULTS_REQUEST';
export const FETCH_MAP_SEARCH_RESULTS_SUCCESS = 'FETCH_MAP_SEARCH_RESULTS_SUCCESS';
export const FETCH_MAP_SEARCH_RESULTS_FAILURE = 'FETCH_MAP_SEARCH_RESULTS_FAILURE';
export const FETCH_SEARCH_RESULTS_CATEGORY = 'FETCH_SEARCH_RESULTS_CATEGORY';
export const FETCH_SEARCH_RESULTS_BY_QUERY = 'FETCH_SEARCH_RESULTS_BY_QUERY';
export const SHOW_SEARCH_RESULTS = 'SHOW_SEARCH_RESULTS';
export const FETCH_SEARCH_RESULTS_BY_LOCATION = 'FETCH_SEARCH_RESULTS_BY_LOCATION';

export const getSearch = (state) => state.search;
export const getMapResultsByLocation = (state) => get(state, 'search.mapSearchResultsByLocation', []);

export const isSearchActive = createSelector(getSearch, (geoSearch) => (
  geoSearch && geoSearch.location && geoSearch.location.length
));

const initialState = {
  mapSearchResultsByLocation: [],
  isLoading: false
};

export const SEARCH_VIEW = {
  MAP_SEARCH: 'MAP_SEARCH',
  SEARCH: 'SEARCH'
};

export default function MapSearchResultsReducer(state = initialState, action) {
  switch (action.type) {
    case routing.dataSearch.type: {
      return {
        ...state,
        query: get(action, 'meta.query.zoekterm')
      };
    }
    case routing.addresses.type: {
      const { query = {} } = action.meta;
      if (Object.prototype.hasOwnProperty.call(query, 'kaart')) {
        return {
          ...state,
          view: SEARCH_VIEW.MAP_SEARCH
        };
      }
      return {
        ...state,
        view: SEARCH_VIEW.SEARCH
      };
    }
    case FETCH_SEARCH_RESULTS_CATEGORY:
      return isObject(state) ? {
        ...state,
        isLoading: true,
        category: action.payload,
        numberOfResults: null
      } : state;

    case FETCH_SEARCH_RESULTS_BY_LOCATION:
      return {
        isLoading: true,
        isFullscreen: false,
        query: null,
        location: BaseCoder.toPrecision(action.payload, 7),
        category: null,
        numberOfResults: null
      };

    case FETCH_SEARCH_RESULTS_BY_QUERY:
      return {
        isLoading: true,
        isFullscreen: true,
        query: action.payload || null,
        location: null,
        category: null,
        numberOfResults: null
      };

    case SHOW_SEARCH_RESULTS:
      return {
        ...state,
        isLoading: false,
        numberOfResults: action.payload
      };
    case FETCH_MAP_SEARCH_RESULTS_REQUEST:
      return {
        ...state,
        isLoading: true
      };

    case FETCH_MAP_SEARCH_RESULTS_SUCCESS: {
      return {
        ...state,
        isLoading: false,
        mapSearchResultsByLocation: action.mapSearchResults
      };
    }

    case FETCH_MAP_SEARCH_RESULTS_FAILURE:
      return { ...state, isLoading: false };

    default:
      return state;
  }
}

export const isSearchLoading = (state) => state[REDUCER_KEY].isLoading;
export const getSearchView = (state) => state[REDUCER_KEY].view;
export const getSearchQuery = (state) => state[REDUCER_KEY].query;
export const getSearchCategory = (state) => state[REDUCER_KEY].category;
export const getNumberOfResults = (state) => state[REDUCER_KEY].numberOfResults;

export const getMapSearchResults = (location, user) => ({
  type: FETCH_MAP_SEARCH_RESULTS_REQUEST,
  location,
  user
});
