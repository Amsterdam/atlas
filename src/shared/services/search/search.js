// Todo: fix / add tests
/* istanbul ignore file */
import SEARCH_CONFIG from './search-config';
import { getByUrl } from '../api/api';
import SHARED_CONFIG from '../shared-config/shared-config';
import { formatCategories, formatCategory, formatLinks } from './search-formatter';

function isString(value) {
  return typeof value === 'string';
}

// Todo: user.scopes.includes(endpoint.authScope)
export function search(query, categorySlug, user) {
  const queries = [];
  const params = { q: query };
  SEARCH_CONFIG.QUERY_ENDPOINTS.forEach((endpoint) => {
    if ((!isString(categorySlug) || categorySlug === endpoint.slug) &&
      endpoint.uri &&
      (!endpoint.authScope || user.scopes.includes(endpoint.authScope))
    ) {
      const options = endpoint.options || {};
      queries.push(
        getByUrl(
          `${SHARED_CONFIG.API_ROOT}${endpoint.uri}`,
          { ...params, ...options }
        ).then((data) => data, () => [])
      );
    }
  });

  if (isString(categorySlug)) {
    // A single category
    return Promise.all(queries).then((searchResults) =>
      [formatCategory(categorySlug, searchResults[0])]
    );
  }
  // All search results
  return Promise.all(queries).then((results) => formatCategories(results, user));
}

export function loadMore(category) {
  return getByUrl(category.next)
    .then((nextPageData) => {
      // Don't change the input, create a new variable
      const output = {};

      output.slug = category.slug;
      output.count = nextPageData.count;
      output.results = category.results.concat(
        formatLinks(category.slug, nextPageData.results)
      );

      if (output.count > output.results.length) {
        output.next = nextPageData._links.next.href;
      } else {
        output.next = null;
      }

      return output;
    });
}

// @TODO remove the exception when backend uses correct sub type name tg-3551
export function replaceBuurtcombinatie(searchResults) {
  const results = [...searchResults];
  results.forEach((result) => {
    result.results.forEach((item) => {
      if (item.subtype === 'buurtcombinatie') {
        item.subtypeLabel = 'wijk'; // eslint-disable-line no-param-reassign
      }
    });
  });

  return results;
}

export function hasLoadMore(category, searchResults, isLoadMoreLoading) {
  return isString(category) &&
    searchResults[0].count > searchResults[0].results.length &&
    !isLoadMoreLoading;
}

export function getNumberOfResults(searchResults) {
  return searchResults.reduce((previous, current) => (
    previous + current.count +
    (current.subResults
      ? current.subResults.reduce((subPrevious, subCurrent) => (
        subPrevious + subCurrent.count
      ), 0)
      : 0)
  ), 0);
}
