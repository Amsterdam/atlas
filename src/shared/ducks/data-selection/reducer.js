import {
  FETCH_DATA_SELECTION_FAILURE,
  FETCH_DATA_SELECTION_REQUEST,
  FETCH_DATA_SELECTION_SUCCESS,
  initialState, ROUTE_DATASET_MAPPER,
  SET_DATASET,
  REMOVE_GEOMETRY_FILTER,
  SET_GEOMETRY_FILTER,
  SET_MARKERS,
  SET_PAGE,
  SET_VIEW,
  REDUCER_KEY
} from './constants';
import { routing } from '../../../app/routes';
import { SET_SELECTION } from '../selection/selection';
import { FETCH_MAP_DETAIL_SUCCESS } from '../../../map/ducks/detail/constants';
import paramsRegistry from '../../../store/params-registry';

export { REDUCER_KEY as DATA_SELECTION };

export default function reducer(state = initialState, action) {
  const enrichedState = {
    ...state,
    ...paramsRegistry.getStateFromQueries(REDUCER_KEY, action),
    ...(ROUTE_DATASET_MAPPER[action.type]) ? { dataset: ROUTE_DATASET_MAPPER[action.type] } : {}
  };

  switch (action.type) {
    case routing.home.type:
    case FETCH_MAP_DETAIL_SUCCESS:
    case SET_SELECTION: {
      return {
        ...initialState
      };
    }

    case FETCH_DATA_SELECTION_REQUEST:
      return {
        ...enrichedState,
        isLoading: true,
        markers: []
      };

    case FETCH_DATA_SELECTION_SUCCESS: {
      return {
        ...enrichedState,
        isLoading: false,
        markers: [],
        errorMessage: '',
        authError: false,
        ...action.payload
      };
    }

    case FETCH_DATA_SELECTION_FAILURE:
      return {
        ...enrichedState,
        isLoading: false,
        authError: (action.payload.error === 'Unauthorized'),
        errorMessage: action.payload.error,
        dataset: action.payload.dataset,
        result: {},
        markers: []
      };

    case SET_MARKERS:
      return {
        ...enrichedState,
        isLoading: false,
        markers: action.payload
      };

    case SET_DATASET:
      return {
        ...enrichedState,
        dataset: action.payload
      };

    case SET_GEOMETRY_FILTER:
      return {
        ...enrichedState,
        geometryFilter: action.payload
      };

    case REMOVE_GEOMETRY_FILTER:
      return {
        ...enrichedState,
        geometryFilter: {}
      };

    case SET_PAGE:
      return {
        ...enrichedState,
        page: action.payload
      };

    case SET_VIEW:
      return {
        ...enrichedState,
        view: action.payload
      };

    default:
      return enrichedState;
  }
}
