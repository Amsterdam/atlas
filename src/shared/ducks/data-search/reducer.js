import {
  FETCH_MAP_SEARCH_RESULTS_FAILURE,
  FETCH_MAP_SEARCH_RESULTS_REQUEST,
  FETCH_MAP_SEARCH_RESULTS_SUCCESS_LIST,
  FETCH_MAP_SEARCH_RESULTS_SUCCESS_PANEL,
  FETCH_QUERY_SEARCH_MORE_RESULTS_SUCCESS,
  FETCH_QUERY_SEARCH_RESULTS_FAILURE,
  FETCH_QUERY_SEARCH_RESULTS_REQUEST,
  FETCH_QUERY_SEARCH_RESULTS_SUCCESS,
  initialState,
  REDUCER_KEY,
  SET_GEO_LOCATION,
  SET_QUERY_CATEGORY
} from './constants';
import { getStateFromQuery } from '../../../store/query-synchronization';
import urlParams from './query';
import { routing } from '../../../app/routes';
import { FETCH_DATA_SELECTION_REQUEST } from '../data-selection/constants';

export { REDUCER_KEY };

export default function reducer(state = initialState, action) {
  const enrichedState = {
    ...state,
    ...getStateFromQuery(urlParams, action)
  };

  switch (action.type) {
    case routing.dataSearchCategory.type:
      return {
        ...enrichedState,
        category: action.payload.category
      };
    case FETCH_QUERY_SEARCH_RESULTS_REQUEST:
      return {
        ...initialState,
        isLoading: true,
        query: action.payload
      };

    case FETCH_QUERY_SEARCH_RESULTS_SUCCESS: {
      const { results, numberOfResults } = action.payload;
      return {
        ...enrichedState,
        isLoading: false,
        numberOfResults,
        resultsQuery: results
      };
    }

    case FETCH_QUERY_SEARCH_MORE_RESULTS_SUCCESS: {
      return {
        ...enrichedState,
        isLoading: false,
        numberOfResults: 999,
        resultsQuery: [{
          ...state.resultsQuery[0],
          ...action.payload[0]
        }]
      };
    }

    case FETCH_MAP_SEARCH_RESULTS_REQUEST:
      return {
        ...initialState,
        isLoading: true,
        geoSearch: action.payload
      };

    case FETCH_MAP_SEARCH_RESULTS_SUCCESS_PANEL: {
      const { results, numberOfResults } = action.payload;
      return {
        ...enrichedState,
        isLoading: false,
        numberOfResults,
        resultsMapPanel: results
      };
    }

    case FETCH_MAP_SEARCH_RESULTS_SUCCESS_LIST: {
      const { results, numberOfResults } = action.payload;
      return {
        ...enrichedState,
        isLoading: false,
        numberOfResults,
        resultsMap: results
      };
    }

    case FETCH_QUERY_SEARCH_RESULTS_FAILURE:
    case FETCH_MAP_SEARCH_RESULTS_FAILURE:
      return {
        ...enrichedState,
        isLoading: false,
        error: action.payload
      };

    case SET_QUERY_CATEGORY:
      return {
        ...enrichedState,
        category: action.payload
      };

    case SET_GEO_LOCATION:
      return {
        ...enrichedState,
        geoSearch: action.payload
      };

    case FETCH_DATA_SELECTION_REQUEST:
      return {
        ...initialState
      };

    default:
      return enrichedState;
  }
}
