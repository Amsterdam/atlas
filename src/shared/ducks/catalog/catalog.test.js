import { takeLatest, put } from 'redux-saga/effects';
import reducer, { fetchCatalogData, watchCatalogList } from './catalog';
import { routing } from '../../../app/routes';
import { fetchDataSelection } from '../../../header/ducks/search/search';

describe('Catalog Reducer', () => {
  it('should set the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should set the detail id from a route change', () => {
    const id = '123foo';
    const result = reducer({}, { type: routing.catalogusDetail.type, payload: { id } });
    expect(result.detail).toBe(id);
  });

  it('should dispatch fetch data on route change', () => {
    const watchGen = watchCatalogList();
    expect(watchGen.next().value).toEqual(
      takeLatest([
        routing.catalogus.type,
        routing.searchCatalog.type
      ], fetchCatalogData)
    );

    const fetchGen = fetchCatalogData();
    expect(fetchGen.next().value).toEqual(
      put(
        fetchDataSelection({
          dataset: 'dcatd',
          view: 'CATALOG',
          page: 1
        })
      )
    );
  });
});
