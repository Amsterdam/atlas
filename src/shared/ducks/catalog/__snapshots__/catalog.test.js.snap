// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog Reducer should set the initial state 1`] = `
Object {
  "endpoint": undefined,
}
`;
